crypto/cloudkms/signer.go

```diff
@@ -64,77 +70,62 @@ func (c *Client) SignerForKey(
 //
 // Reference: https://cloud.google.com/kms/docs/create-validate-signatures
 func (s *Signer) Sign(message []byte) ([]byte, error) {
-	hasher, err := crypto.NewHasher(s.hashAlgo)
-	if err != nil {
-		return nil, fmt.Errorf("cloudkms: failed to instantiate hasher: %w", err)
-	}
-	digest := hasher.ComputeHash(message)
-
-	digestMessage, err := makeDigest(s.hashAlgo, digest)
-	if err != nil {
-		return nil, fmt.Errorf("cloudkms: failed to construct digest: %w", err)
-	}
 
 	request := &kmspb.AsymmetricSignRequest{
-		Name:   s.key.ResourceID(),
-		Digest: digestMessage,
+		Name:       s.key.ResourceID(),
+		Data:       message,
+		DataCrc32C: checksum(message),
 	}
-
 	result, err := s.client.AsymmetricSign(s.ctx, request)
 	if err != nil {
 		return nil, fmt.Errorf("cloudkms: failed to sign: %w", err)
 	}
-
-	sig, err := parseSignature(result.Signature)
+	sig, err := parseSignature(result.Signature, s.curve)
 	if err != nil {
 		return nil, fmt.Errorf("cloudkms: failed to parse signature: %w", err)
 	}
-
 	return sig, nil
 }
 
-func makeDigest(hashAlgo crypto.HashAlgorithm, digest []byte) (*kmspb.Digest, error) {
-	switch hashAlgo {
-	case crypto.SHA2_256:
-		return &kmspb.Digest{Digest: &kmspb.Digest_Sha256{Sha256: digest}}, nil
-	case crypto.SHA2_384:
-		return &kmspb.Digest{Digest: &kmspb.Digest_Sha384{Sha384: digest}}, nil
-	}
-
-	return nil, fmt.Errorf("unsupported hash algorithm %s", hashAlgo)
+func checksum(data []byte) *wrapperspb.Int64Value {
+	// compute CRC32
+	table := crc32.MakeTable(crc32.Castagnoli)
+	checksum := crc32.Checksum(data, table)
+	val := wrapperspb.Int64(int64(checksum))
+	return val
 }
 
-// ecCoupleComponentSize is size of a component in either (r,s) couple for an elliptical curve signature
-// or (x,y) identifying a public key. Component size is needed for encoding couples comprised of variable length
-// numbers to []byte encoding. They are not always the same length, so occasionally padding is required.
-// Here's how one calculates the required length of each component:
-// 		ECDSA_CurveBits = 256
-// 		ecCoupleComponentSize := ECDSA_CurveBits / 8
-// 		if ECDSA_CurveBits % 8 > 0 {
-//			ecCoupleComponentSize++
-// 		}
-const ecCoupleComponentSize = 32
-
-func parseSignature(signature []byte) ([]byte, error) {
+// parseSignature parses an asn1 stucture (R,S) into a slice of bytes as required by the `Siger.Sign` method.
+func parseSignature(kmsSignature []byte, curve crypto.SignatureAlgorithm) ([]byte, error) {
 	var parsedSig struct{ R, S *big.Int }
-	if _, err := asn1.Unmarshal(signature, &parsedSig); err != nil {
+	if _, err := asn1.Unmarshal(kmsSignature, &parsedSig); err != nil {
 		return nil, fmt.Errorf("asn1.Unmarshal: %w", err)
 	}
 
-	rBytes := parsedSig.R.Bytes()
-	rBytesPadded := rightPad(rBytes, ecCoupleComponentSize)
+	curveOrderLen := curveOrder(curve)
+	signature := make([]byte, 2*curveOrderLen)
 
+	// left pad R and S with zeroes
```

Comment:
So just to confirm: we want to left pad instead of right pad?

Reply:
if `(b_0, b_1, ..., b_n)` is the big endian representation of `R`, what we want to get is `[ 0, 0, .. 0, b_0, b_1, .., b_n]`. I assumed this is called left padding. Btw, the original code did it right. 

What I updated is to use the curve order length to pad, instead of the "curve size" or prime length (that is the 256 bits in P256 for instance). These are 2 different values, but they happened to be the same for P256. For a general curve, `R` and `S` should use the curve order.